File: DSAlgorithms/Leetcode/ts/src/arrays_hash/438-all_anagrams_in_string.ts
const map_equality = (map1: Map<string, number>, map2: Map<string, number>): boolean => {
    let return_value: boolean = true;
    map1.forEach((value, key) => {
       if(value !== map2.get(key))
            return_value = false;
    })
    return return_value;
}
const findAnagrams = (s: string, p: string): number[] =>{
    const result: number[] = [];
    if(p.length > s.length)
        return result;
    const p_hash = new Map<string, number>();
    const s_hash = new Map<string, number>();
    for(let i = 0; i < p.length; i++)
    {
        if(p_hash.has(p[i]))
        {
            const current_val: number = p_hash.get(p[i]);
            p_hash.set(p[i], current_val + 1);
        }
        else 
            p_hash.set(p[i], 1);
        if(s_hash.has(s[i]))
        {
            const current_val: number = s_hash.get(s[i]);
            s_hash.set(s[i], current_val + 1);
        }
        else 
            s_hash.set(s[i], 1);
    }
    if(map_equality(s_hash, p_hash))
    {
        console.log("they are eq");
        result.push(0);
    }
    let left: number = 0;
    for(let right = p.length; right < s.length; right++)
    {
        if(s_hash.has(s[right]))
        {
            const current_val: number = s_hash.get(s[right]);
            s_hash.set(s[right], current_val + 1);
        }
        else 
            s_hash.set(s[right], 1);

        if(s_hash.has(s[left]))
        {
            const current_val: number = s_hash.get(s[left]);
            s_hash.set(s[left], current_val - 1);
        }
        if(s_hash.get(s[left]) === 0)
            s_hash.delete(s[left]);
        left ++;
        if(map_equality(s_hash, p_hash))
            result.push(left);
    }
    return result;
};

File: DSAlgorithms/Leetcode/c++/arrays_hash/438-all_anagrams_in_string.cpp
class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        std::vector<int> result;
        if(p.size() > s.size())
            return result;
        std::unordered_map<int, int> p_hash;
        std::unordered_map<int, int> s_hash;
        for(int i = 0; i < p.size(); i++)
        {
            p_hash[p[i]]++; //add all p chars to hash
            s_hash[s[i]]++; //add p.size chars of s to hash
        }
            if(p_hash == s_hash)
                result.push_back(0);
        int left = 0;
        for(int right = p.size(); right < s.size(); right++)
        {
            s_hash[s[right]]++;
            s_hash[s[left]]--;
            if(s_hash[s[left]] == 0)
                s_hash.erase(s[left]);
            left ++;
            /*
            equal_hashes = true;
            for(auto [key, value]: p_hash)
            {
                if(value != s_hash[key])
                    {
                        equal_hashes = false;
                        break;
                    }
            }
            if(equal_hashes)
                result.push_back(left);
            */
            if(p_hash == s_hash)
                result.push_back(left);
        }
        return result;
    }

};